cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,869 @@
+         * May transition back to:
+         * <dl>
+         *     <dt>PendingProposal</dt><dd>If the user sets annotation strimzi.io/rebalance=refresh but the proposal is not ready yet.</dd>
+         *     <dt>ProposalReady</dt><dd>If the user sets annotation strimzi.io/rebalance=refresh and the proposal is already ready.</dd>
+         * </dl>
+         */
+        Stopped,
+        /**
+         * There's been some error.
+         * Transitions to:
+         * <dl>
+         *     <dt>New</dt><dd>If the error was caused by the resource itself that was fixed by the user.</dd>
+         * </dl>
+         */
+        NotReady,
+        /**
+         * The rebalance is complete and there is no transition from this state.
+         * The resource is eligible for garbage collection after a configurable delay.
+         * There is no transition from this state to a new one.
+         */
+        Ready
+    }
+
+    enum RebalanceAnnotation {
+        /**
+         * No annotation set on the rebalance resource.
+         */
+        none,
+        /**
+         * Used to approve a rebalance proposal and trigger the actual rebalancing.
+         * This value should only be used when in the {@code ProposalReady} state.
+         */
+        approve,
+        /**
+         * Used to stop a request for an actual ongoing rebalancing.
+         * This value should only be used when in the {@code Rebalancing} state.
+         */
+        stop,
+        /**
+         * Used to refresh a ready rebalance proposal or to restart a stopped request for getting a rebalance proposal.
+         * This value should only be used when in the {@code ProposalReady} or {@code Stopped} states.
+         */
+        refresh,
+        /**
+         * Any other unsupported/unknown annotation value.
+         */
+        unknown
+    }
+
+    private Future<Void> reconcile(Reconciliation reconciliation, String host, CruiseControlApi apiClient, KafkaRebalance clusterRebalance, State currentState, RebalanceAnnotation rebalanceAnnotation) {
+        RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder = new RebalanceOptions.RebalanceOptionsBuilder();
+        if (clusterRebalance.getSpec().getGoals() != null) {
+            rebalanceOptionsBuilder.withGoals(clusterRebalance.getSpec().getGoals());
+        }
+        if (clusterRebalance.getSpec().isSkipHardGoalCheck()) {
+            rebalanceOptionsBuilder.withSkipHardGoalCheck();
+        }
+        log.info("{}: Rebalance action from state [{}]", reconciliation, currentState);
+
+        return computeNextStatus(reconciliation, host, apiClient, clusterRebalance, currentState, rebalanceAnnotation, rebalanceOptionsBuilder)
+                .compose(desiredStatus -> {
+                    // due to a long rebalancing operation that takes the lock for the entire period, more events related to resource modification could be
```

Comment:
Wait, this sounds confusing ... this for sure does nto work synchronously but is checking the states with Cruise Control asynchronously at every reconciliation, or? So what entire period does the lock need to be hold?

Reply:
When a rebalance resource is created, during the pending proposal computation and rebalancing a lock is hold on the resource with a timer/polling mechanism to check the action status on the CC REST API. It's released when the proposal is ready or the rebalancing is end. I did a lot of back and forth with @tombentley 3 months ago about doing this across reconciliation or in one reconciliation holding the lock. I should get our discussion but this came as the best solution. Even in this case, changing could mean starting from scratch.

Reply:
I don't really understand this. You have periodic reconciliation -> when it is started you check the CC API and update the resource. Why is there any need for an additional timer?

Reply:
What is the impact on resource consumption of this? We have already problem with this and I can immagine when many more addiitonal recocnile loops are added by this it will further increase?

Reply:
the first bad thing I remember is that you should wait for 2 minutes to have the resource updated but it's just a minor thing. The discussion was long over Christmas, my memory is bad and I should get the discussion content if possible with @tombentley :-)

Reply:
Not sure how the 2 minutes are a problem when the rebalance might AFAIU take hours or days. Anyway, probably too late to change now.

Reply:
* How long the proposal takes will depend on many factors. I don't think we really have enough experience of the parameter space to have any intuition, but I think days is unlikely. Seconds, minutes and hours all seem likely.
* But using a 2nd timer we are, in effect, optimising the <2 minutes case. We can form a judgement about whether this is the right way to do it when we have more experience.
* Whether a 2nd timer takes up much resources is something else where we have no data. 

So I think it's better to keep it as it is and be open to revising this later when we have more of a feeling for these things. Another option would be to add some websocket notification endpoint to CC so it could tell us without our needing to poll.

Reply:
You really thing that majority of rebalances take less than 2 minutes? If I follow the idea that we do not have the data, then the simplest solution should have been chosen first and that is using the regular reocnciliation loop. I think that if nothing else this would cause lot of confusion around the locks.

Reply:
The "simplest" solution (as you call it) was tried and it didn't work well. Tbh it was 3 months ago and I don't remember the reason. Maybe it was my fault maybe not. Right now I think that this works and we can live with it, exploring and trying to use the reconcile loop with the next iteration in a future release (after 0.18.0).